drivers/gpio/gpio_sam0.c

```diff
@@ -299,86 +299,64 @@ static const struct gpio_driver_api gpio_sam0_api = {
 
 static int gpio_sam0_init(struct device *dev) { return 0; }
 
-/* Port A */
-#if DT_ATMEL_SAM0_GPIO_PORT_A_BASE_ADDRESS
-
-static const struct gpio_sam0_config gpio_sam0_config_0 = {
-	.common = {
-		.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_NGPIOS(DT_INST_0_ATMEL_SAM0_GPIO_NGPIOS),
-	},
-	.regs = (PortGroup *)DT_ATMEL_SAM0_GPIO_PORT_A_BASE_ADDRESS,
 #ifdef CONFIG_SAM0_EIC
-	.id = 0,
-#endif
-};
-
-static struct gpio_sam0_data gpio_sam0_data_0;
-
-DEVICE_AND_API_INIT(gpio_sam0_0, DT_ATMEL_SAM0_GPIO_PORT_A_LABEL,
-		    gpio_sam0_init, &gpio_sam0_data_0, &gpio_sam0_config_0,
-		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
-		    &gpio_sam0_api);
-#endif
-
-/* Port B */
-#if DT_ATMEL_SAM0_GPIO_PORT_B_BASE_ADDRESS
-
-static const struct gpio_sam0_config gpio_sam0_config_1 = {
-	.common = {
-		.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_NGPIOS(DT_INST_1_ATMEL_SAM0_GPIO_NGPIOS),
-	},
-	.regs = (PortGroup *)DT_ATMEL_SAM0_GPIO_PORT_B_BASE_ADDRESS,
-#ifdef CONFIG_SAM0_EIC
-	.id = 1,
+#define GPIO_SAM0_DEVICE_CONFIG(n)					  \
+static const struct gpio_sam0_config gpio_sam0_config_##n = {		  \
+	.common = {							  \
+		.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_NGPIOS(	  \
+			DT_INST_##n##_ATMEL_SAM0_GPIO_NGPIOS),		  \
+	},								  \
+	.regs = (PortGroup *) DT_INST_##n##_ATMEL_SAM0_GPIO_BASE_ADDRESS, \
+	.id = n,							  \
```

Comment:
Looking at the implementation of `sam0_eic_acquire` -> `sam0_eic_map_to_line` function it seems like the assignment of the id to the gpio instance is strict, i.e. id 0 corresponds to Port A.

Reply:
This was not the first case. I've been solving following #23335 proposal and then with b36c80c it's possible solve that kind of issue.